shared/src/commonMain/kotlin/co/touchlab/kampkit/models/DataState.kt

```diff
@@ -1,8 +1,24 @@
 package co.touchlab.kampkit.models
 
-data class DataState<out T>(
-    val data: T? = null,
-    val exception: String? = null,
-    val empty: Boolean = false,
-    val loading: Boolean = false
-)
+sealed class DataState<out T>(open val loading: Boolean) {
+    class Success<T>(val data: T, override val loading: Boolean = false) : DataState<T>(loading) {
+        override fun copy(isLoading: Boolean): DataState<T> = Success(data, isLoading)
+        override fun equals(other: Any?) = other is Success<*> && other.data == data && other.loading == loading
+    }
+
+    class Error(val exception: String, override val loading: Boolean = false) : DataState<Nothing>(loading) {
+        override fun copy(isLoading: Boolean): DataState<Nothing> = Error(exception, isLoading)
+        override fun equals(other: Any?) = other is Error && other.exception == exception && other.loading == loading
+    }
+
+    class Empty(override val loading: Boolean = false) : DataState<Nothing>(loading) {
+        override fun copy(isLoading: Boolean) = this
+        override fun equals(other: Any?) = other is Empty && other.loading == loading
+    }
+
+    object Loading : DataState<Nothing>(true) {
+        override fun copy(isLoading: Boolean): DataState<Nothing> = this
+    }
+
+    abstract fun copy(isLoading: Boolean): DataState<T>
```

Comment:
It feels unintuitive to me that for some states, `copy()` ignores the parameter and is a no-op. Here's some alternatives I can think of:
1. Only define it on classes where it's meaningful. This means you can probably return to using the data class copy. Downside is this means you need to cast before you can call it.
2. Remove `Loading` state and just have `Success`, `Error`, and `Empty` with a loading boolean. Downside is you might not want to display the `Empty` state content underneath the loading spinner on the first load
3. Have `Empty` potentially return a `Loading` and vice-versa, depending on the parameter. Downside is this might make the generics messier and gets further away from what a developer will expect from a `copy()` method.